File: services/conversation/conversation.go
package conversation

import (
	"context"
	"time"

	"github.com/0xdeafcafe/bloefish/libraries/cher"
)

type Service interface {
	CreateConversation(ctx context.Context, req *CreateConversationRequest) (*CreateConversationResponse, error)
	CreateConversationMessage(ctx context.Context, req *CreateConversationMessageRequest) (*CreateConversationMessageResponse, error)
	GetInteraction(ctx context.Context, req *GetInteractionRequest) (*GetInteractionResponse, error)
	GetConversationWithInteractions(ctx context.Context, req *GetConversationWithInteractionsRequest) (*GetConversationWithInteractionsResponse, error)
	ListConversationsWithInteractions(ctx context.Context, req *ListConversationsWithInteractionsRequest) (*ListConversationsWithInteractionsResponse, error)
	DeleteConversations(ctx context.Context, req *DeleteConversationsRequest) error
	DeleteInteractions(ctx context.Context, req *DeleteInteractionsRequest) error
	UpdateInteractionExcludedState(ctx context.Context, req *UpdateInteractionExcludedStateRequest) error
}

type ActorType string

const (
	ActorTypeUser ActorType = "user"
	ActorTypeBot  ActorType = "bot"
)

type Actor struct {
	Type       ActorType `json:"type"`
	Identifier string    `json:"identifier"`
}

type AIRelayOptions struct {
	ProviderID string `json:"provider_id"`
	ModelID    string `json:"model_id"`
}

type CreateConversationRequest struct {
	IdempotencyKey string          `json:"idempotency_key"`
	Owner          *Actor          `json:"owner"`
	AIRelayOptions *AIRelayOptions `json:"ai_relay_options"`
}

type CreateConversationResponse struct {
	ID             string          `json:"id"`
	Owner          *Actor          `json:"owner"`
	AIRelayOptions *AIRelayOptions `json:"ai_relay_options"`

	Title           *string `json:"title"`
	StreamChannelID string  `json:"stream_channel_id"`

	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
	DeletedAt *time.Time `json:"deleted_at"`
}

type CreateConversationMessageRequest struct {
	ConversationID string                                   `json:"conversation_id"`
	IdempotencyKey string                                   `json:"idempotency_key"`
	MessageContent string                                   `json:"message_content"`
	FileIDs        []string                                 `json:"file_ids"`
	Owner          *Actor                                   `json:"owner"`
	AIRelayOptions *AIRelayOptions                          `json:"ai_relay_options"`
	Options        *CreateConversationMessageRequestOptions `json:"options"`
}

type CreateConversationMessageRequestOptions struct {
	UseStreaming bool `json:"use_streaming"`
}

type CreateConversationMessageResponse struct {
	ConversationID      string                                        `json:"conversation_id"`
	InputInteraction    *CreateConversationMessageResponseInteraction `json:"input_interaction"`
	ResponseInteraction *CreateConversationMessageResponseInteraction `json:"response_interaction"`
	StreamChannelID     string                                        `json:"stream_channel_id"`
}

type CreateConversationMessageResponseInteraction struct {
	ID              string   `json:"id"`
	FileIDs         []string `json:"file_ids"`
	StreamChannelID string   `json:"stream_channel_id"`

	MarkedAsExcludedAt *time.Time `json:"marked_as_excluded_at"`

	MessageContent string   `json:"message_content"`
	Errors         []cher.E `json:"errors"`

	Owner          *Actor          `json:"owner"`
	AIRelayOptions *AIRelayOptions `json:"ai_relay_options"`

	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	DeletedAt   *time.Time `json:"deleted_at"`
	CompletedAt *time.Time `json:"completed_at"`
}

type GetInteractionRequest struct {
	InteractionID string `json:"interaction_id"`
}

type GetInteractionResponse struct {
	ID             string   `json:"id"`
	ConversationID string   `json:"conversation_id"`
	FileIDs        []string `json:"file_ids"`

	MarkedAsExcludedAt *time.Time `json:"marked_as_excluded_at"`

	MessageContent string   `json:"message_content"`
	Errors         []cher.E `json:"errors"`

	Owner          *Actor          `json:"owner"`
	AIRelayOptions *AIRelayOptions `json:"ai_relay_options"`

	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	DeletedAt   *time.Time `json:"deleted_at"`
	CompletedAt *time.Time `json:"completed_at"`
}

type GetConversationWithInteractionsRequest struct {
	ConversationID string `json:"conversation_id"`
}

type GetConversationWithInteractionsResponse struct {
	ConversationID string          `json:"conversation_id"`
	Owner          *Actor          `json:"owner"`
	AIRelayOptions *AIRelayOptions `json:"ai_relay_options"`

	Title           *string `json:"title"`
	StreamChannelID string  `json:"stream_channel_id"`

	Interactions []*GetConversationWithInteractionsResponseInteraction `json:"interactions"`

	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
	DeletedAt *time.Time `json:"deleted_at"`
}

type GetConversationWithInteractionsResponseInteraction struct {
	ID              string   `json:"id"`
	FileIDs         []string `json:"file_ids"`
	StreamChannelID string   `json:"stream_channel_id"`

	MarkedAsExcludedAt *time.Time `json:"marked_as_excluded_at"`

	MessageContent string   `json:"message_content"`
	Errors         []cher.E `json:"errors"`

	Owner          *Actor          `json:"owner"`
	AIRelayOptions *AIRelayOptions `json:"ai_relay_options"`

	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	CompletedAt *time.Time `json:"completed_at"`
	DeletedAt   *time.Time `json:"deleted_at"`
}

type ListConversationsWithInteractionsRequest struct {
	Owner *Actor `json:"owner"`
}

type ListConversationsWithInteractionsResponse struct {
	Conversations []*ListConversationsWithInteractionsResponseConversation `json:"conversations"`
}

type ListConversationsWithInteractionsResponseConversation struct {
	ID             string          `json:"id"`
	Owner          *Actor          `json:"owner"`
	AIRelayOptions *AIRelayOptions `json:"ai_relay_options"`

	Title           *string `json:"title"`
	StreamChannelID string  `json:"stream_channel_id"`

	Interactions []*ListConversationsWithInteractionsResponseConversationInteraction `json:"interactions"`

	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
	DeletedAt *time.Time `json:"deleted_at"`
}

type ListConversationsWithInteractionsResponseConversationInteraction struct {
	ID              string   `json:"id"`
	FileIDs         []string `json:"file_ids"`
	StreamChannelID string   `json:"stream_channel_id"`

	MarkedAsExcludedAt *time.Time `json:"marked_as_excluded_at"`

	MessageContent string   `json:"message_content"`
	Errors         []cher.E `json:"errors"`

	Owner          *Actor          `json:"owner"`
	AIRelayOptions *AIRelayOptions `json:"ai_relay_options"`

	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
	CompletedAt *time.Time `json:"completed_at"`
	DeletedAt   *time.Time `json:"deleted_at"`
}

type DeleteConversationsRequest struct {
	ConversationIDs []string `json:"conversation_ids"`
}

type DeleteInteractionsRequest struct {
	InteractionIDs []string `json:"interaction_ids"`
}

type UpdateInteractionExcludedStateRequest struct {
	InteractionID string `json:"interaction_id"`
	Excluded      bool   `json:"excluded"`
}

File: applications/web/src/api/bloefish/conversation.ts
import { createApi } from '@reduxjs/toolkit/query/react';
import type {
	GetConversationWithInteractionsResponse,
	CreateConversationMessageRequest,
	CreateConversationMessageResponse,
	CreateConversationRequest,
	CreateConversationResponse,
	GetConversationWithInteractionsRequest,
	ListConversationsWithInteractionsRequest,
	ListConversationsWithInteractionsResponse,
	DeleteConversationsRequest,
	DeleteInteractionsRequest,
	UpdateInteractionExcludedStateRequest,
} from './conversation.types';
import { createBaseQueryWithSnake } from './base';
import { deleteConversations, deleteInteractions, injectConversations, updateInteractionIncludedState } from '~/features/conversations/store';

export const conversationApi = createApi({
	reducerPath: 'api.bloefish.conversation',
	baseQuery: createBaseQueryWithSnake('http://svc_conversation.bloefish.local:4002/rpc/'),

	endpoints: (builder) => ({
		createConversation: builder.mutation<CreateConversationResponse, CreateConversationRequest>({
			query: (body) => ({
				url: '2025-02-12/create_conversation',
				body,
			}),
		}),

		createConversationMessage: builder.mutation<CreateConversationMessageResponse, CreateConversationMessageRequest>({
			query: (body) => ({
				url: '2025-02-12/create_conversation_message',
				body,
			}),
		}),

		getConversationWithInteractions: builder.query<GetConversationWithInteractionsResponse, GetConversationWithInteractionsRequest>({
			query: (body) => ({
				url: '2025-02-12/get_conversation_with_interactions',
				body,
			}),
			async onQueryStarted(_, { dispatch, queryFulfilled }) {
				try {
					const { data } = await queryFulfilled;

					dispatch(injectConversations([{
						id: data.id,
						owner: data.owner,
						aiRelayOptions: data.aiRelayOptions,

						title: data.title,
						streamChannelId: data.streamChannelId,

						interactions: Object.fromEntries(data.interactions.map((interaction) => [
							interaction.id,
							{
								conversationId: data.id,
								id: interaction.id,
								streamChannelId: interaction.streamChannelId,
								
								markedAsExcludedAt: interaction.markedAsExcludedAt,

								messageContent: interaction.messageContent,
								errors: interaction.errors,

								owner: interaction.owner,
								aiRelayOptions: interaction.aiRelayOptions,

								createdAt: interaction.createdAt,
								updatedAt: interaction.updatedAt,
								deletedAt: interaction.deletedAt,
								completedAt: interaction.completedAt,
							}
						])),

						createdAt: data.createdAt,
						updatedAt: data.updatedAt,
						deletedAt: data.deletedAt,
					}]));
				} catch (error) {
					console.error(error);
				}
			},
		}),

		listConversationsWithInteractions: builder.query<ListConversationsWithInteractionsResponse, ListConversationsWithInteractionsRequest>({
			query: (body) => ({
				url: '2025-02-12/list_conversations_with_interactions',
				body,
			}),
			async onQueryStarted(_, { dispatch, queryFulfilled }) {
				try {
					const { data } = await queryFulfilled;

					dispatch(injectConversations(data.conversations.map((conversation) => ({
						id: conversation.id,
						owner: conversation.owner,
						aiRelayOptions: conversation.aiRelayOptions,

						title: conversation.title,
						streamChannelId: conversation.streamChannelId,

						interactions: Object.fromEntries(conversation.interactions.map((interaction) => [
							interaction.id,
							{
								conversationId: conversation.id,
								id: interaction.id,
								streamChannelId: interaction.streamChannelId,

								markedAsExcludedAt: interaction.markedAsExcludedAt,

								messageContent: interaction.messageContent,
								errors: interaction.errors,

								aiRelayOptions: interaction.aiRelayOptions,
								owner: interaction.owner,

								createdAt: interaction.createdAt,
								updatedAt: interaction.updatedAt,
								completedAt: interaction.completedAt,
								deletedAt: interaction.deletedAt,
							}
						])),

						createdAt: conversation.createdAt,
						updatedAt: conversation.updatedAt,
						deletedAt: conversation.deletedAt,
					}))));
				} catch (error) {
					console.error(error);
				}
			},
		}),

		deleteConversations: builder.mutation<void, DeleteConversationsRequest>({
			query: (body) => ({
				url: '2025-02-12/delete_conversations',
				body,
			}),
			async onQueryStarted(req, { dispatch, queryFulfilled }) {
				try {
					await queryFulfilled;

					dispatch(deleteConversations({
						conversationIds: req.conversationIds,
					}));
				} catch (error) {
					console.error(error);
				}
			},
		}),

		deleteInteractions: builder.mutation<void, DeleteInteractionsRequest>({
			query: (body) => ({
				url: '2025-02-12/delete_interactions',
				body,
			}),
			async onQueryStarted(req, { dispatch, queryFulfilled }) {
				try {
					await queryFulfilled;

					dispatch(deleteInteractions({ interactionIds: req.interactionIds }));
				} catch (error) {
					console.error(error);
				}
			},
		}),

		updateInteractionExcludedState: builder.mutation<void, UpdateInteractionExcludedStateRequest>({
			query: (body) => ({
				url: '2025-02-12/update_interaction_excluded_state',
				body,
			}),
			async onQueryStarted(req, { dispatch, queryFulfilled }) {
				try {
					await queryFulfilled;

					dispatch(updateInteractionIncludedState({
						interactionId: req.interactionId,
						excluded: req.excluded,
					}));
				} catch (error) {
					console.error(error);
				}
			},
		}),
	}),
});
